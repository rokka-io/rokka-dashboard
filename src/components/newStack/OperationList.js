import React from 'react'
import PropTypes from 'prop-types'
import FormGroup from '../forms/FormGroup'
import Options from '../Options'
import Operation from '../operations'

const OperationList = ({
  name,
  options,
  stacks,
  onChangeName,
  onChangeOptions,
  onChangeOperation,
  addOperation,
  removeOperation,
  setActiveOperation,
  onMoveOperation,
  onSelectAddOperation,
  selectedOperation,
  activeOperation = 0,
  addedOperations = [],
  availableOperations = {},
  stackOptions = null
}) => {
  return (
    <div className="col-md-7 col-sm-7">
      <h3 className="rka-h2 mv-md">Stack details</h3>
      <FormGroup label="Name" required>
        <input
          type="text"
          className="rka-input-txt"
          id="name"
          name="name"
          onChange={onChangeName}
          value={name}
        />
      </FormGroup>

      <Options
        defaultOptions={stackOptions ? stackOptions.properties : {}}
        options={options}
        onChange={onChangeOptions}
        stacks={stacks}
      />

      <h3 className="rka-h2 mv-md">Operations</h3>
      {addedOperations.map((operation, index) => {
        return (
          <Operation
            availableOperations={availableOperations}
            key={`operation-${operation.id}-${operation.name}`}
            operation={operation}
            index={index}
            isActive={index === activeOperation}
            onChange={onChangeOperation}
            removeOperation={removeOperation}
            setActiveOperation={setActiveOperation}
            onMoveOperation={onMoveOperation}
          />
        )
      })}

      <div className="pa-md bor-light mt-md">
        <h3 className="rka-h3 mb-md">New operation</h3>
        <div className="rka-form-group">
          <select className="rka-select" onChange={onSelectAddOperation} value={selectedOperation}>
            {Object.keys(availableOperations)
              .filter(name => name !== 'noop')
              .sort()
              .map(name => (
                <option key={name} value={name}>
                  {name}
                </option>
              ))}
          </select>
        </div>
        <button className="rka-button rka-button-brand rka-button-sm" onClick={addOperation}>
          Add operation
        </button>
      </div>
    </div>
  )
}
OperationList.propTypes = {
  name: PropTypes.string.isRequired,
  options: PropTypes.object.isRequired,
  onChangeName: PropTypes.func.isRequired,
  onChangeOptions: PropTypes.func.isRequired,
  onChangeOperation: PropTypes.func.isRequired,
  addOperation: PropTypes.func.isRequired,
  removeOperation: PropTypes.func.isRequired,
  setActiveOperation: PropTypes.func.isRequired,
  onMoveOperation: PropTypes.func.isRequired,
  onSelectAddOperation: PropTypes.func.isRequired,
  selectedOperation: PropTypes.string.isRequired,
  activeOperation: PropTypes.number,
  addedOperations: PropTypes.array,
  availableOperations: PropTypes.object,
  stacks: PropTypes.object,
  stackOptions: PropTypes.shape({
    properties: PropTypes.object.isRequired
  })
}

export default OperationList
