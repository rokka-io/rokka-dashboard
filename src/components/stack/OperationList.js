import React from 'react'
import PropTypes from 'prop-types'
import Operation from '../operations/Operation'

const OperationList = ({
  onChangeOperation,
  onAddOperation,
  onRemoveOperation,
  setActiveOperation,
  onMoveOperation,
  onSelectAddOperation,
  selectedOperation,
  activeOperation = 0,
  addedOperations = [],
  availableOperations = {}
}) => {
  return (
    <>
      <h3 className="rka-h2 mv-md">Operations</h3>
      {addedOperations.map((operation, index) => {
        return (
          <Operation
            availableOperations={availableOperations}
            key={`operation-${operation.id}-${operation.name}`}
            operation={operation}
            index={index}
            isActive={index === activeOperation}
            onChange={onChangeOperation}
            removeOperation={onRemoveOperation}
            setActiveOperation={setActiveOperation}
            onMoveOperation={onMoveOperation}
          />
        )
      })}

      {onAddOperation && (
        <div className="pa-md bor-light mt-md">
          <h3 className="rka-h3 mb-md">New operation</h3>
          <div className="rka-form-group">
            <select
              className="rka-select"
              onChange={onSelectAddOperation}
              value={selectedOperation}
            >
              {Object.keys(availableOperations)
                .filter(name => name !== 'noop')
                .sort()
                .map(name => (
                  <option key={name} value={name}>
                    {name}
                  </option>
                ))}
            </select>
          </div>
          <button className="rka-button rka-button-brand rka-button-sm" onClick={onAddOperation}>
            Add operation
          </button>
        </div>
      )}
    </>
  )
}
OperationList.propTypes = {
  onChangeOperation: PropTypes.func,
  onAddOperation: PropTypes.func,
  onRemoveOperation: PropTypes.func,
  setActiveOperation: PropTypes.func,
  onMoveOperation: PropTypes.func,
  onSelectAddOperation: PropTypes.func,
  selectedOperation: PropTypes.string,
  activeOperation: PropTypes.number,
  addedOperations: PropTypes.array,
  availableOperations: PropTypes.object
}

export default OperationList
