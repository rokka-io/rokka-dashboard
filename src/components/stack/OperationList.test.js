import React, { Component } from 'react'
import { DragDropContext } from 'react-dnd'
import TestBackend from 'react-dnd-test-backend'
import OperationList from './OperationList'
import renderer from 'react-test-renderer'

function wrapInTestContext(DecoratedComponent) {
  class TestContextContainer extends Component {
    render() {
      return <DecoratedComponent {...this.props} />
    }
  }

  return DragDropContext(TestBackend)(TestContextContainer)
}

test('OperationList does render with minimal props', () => {
  const failingFunc = () => {
    throw new Error('Should not be called')
  }
  const props = {
    name: '',
    options: {},
    stacks: {},
    onChangeName: failingFunc,
    onChangeOptions: failingFunc,
    onChangeOperation: failingFunc,
    addOperation: failingFunc,
    removeOperation: failingFunc,
    setActiveOperation: failingFunc,
    onMoveOperation: failingFunc,
    onSelectAddOperation: failingFunc,
    selectedOperation: ''
  }
  const component = renderer.create(<OperationList {...props} />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('OperationList does render with all props', () => {
  const failingFunc = () => {
    throw new Error('Should not be called')
  }
  const props = {
    name: '',
    options: {},
    onChangeName: failingFunc,
    onChangeOptions: failingFunc,
    onChangeOperation: failingFunc,
    onAddOperation: failingFunc,
    onRemoveOperation: failingFunc,
    setActiveOperation: failingFunc,
    onMoveOperation: failingFunc,
    onSelectAddOperation: failingFunc,
    selectedOperation: '',
    activeOperation: 1,
    availableOperations: {
      rotate: {
        properties: {
          angle: {
            minimum: 0,
            maximum: 360,
            default: 90
          },
          background_color: {
            default: '#000000'
          },
          background_opacity: {
            minimum: 0,
            maximum: 100
          }
        }
      },
      crop: {
        properties: {
          width: {
            minimum: 0,
            maximum: 10000
          },
          height: {
            minimum: 0,
            maximum: 10000
          },
          anchor: {
            default: 'left'
          }
        }
      }
    },
    addedOperations: [
      {
        id: '1',
        name: 'rotate',
        options: {},
        errors: {}
      },
      {
        id: '2',
        name: 'crop',
        options: {},
        errors: {}
      }
    ],
    stackOptions: {
      properties: {}
    }
  }
  const WrappedOperationList = wrapInTestContext(OperationList)
  const component = renderer.create(<WrappedOperationList {...props} />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
