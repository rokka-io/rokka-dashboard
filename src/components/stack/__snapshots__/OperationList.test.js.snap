// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OperationList does render with all props 1`] = `
Array [
  <h3
    className="rka-h2 mv-md"
  >
    Operations
  </h3>,
  <div
    className="pa-md mb-xs pos-r bor-light"
  >
    <div>
      <button
        className="rka-close-icon"
        onClick={[Function]}
      />
      <button
        className="rka-edit-icon"
        onClick={[Function]}
      />
      <span
        className="rka-move-icon"
      />
      <h3
        className="rka-h3 txt-cap"
      >
        rotate
      </h3>
    </div>
    <div
      className=""
    >
      <div
        className="rka-operation-content"
      >
        <div
          className="rka-form-group"
        >
          <label
            className="rka-label txt-cap"
            htmlFor="background_color"
          >
            background color
          </label>
          <div
            className="rka-form-group"
          >
            <input
              className="rka-input-txt"
              disabled={false}
              name="background_color"
              onClick={[Function]}
              placeholder="Hex color (default: #000000)"
              readOnly={true}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="rka-form-group"
        >
          <label
            className="rka-label txt-cap"
            htmlFor="angle"
          >
            angle
          </label>
          <input
            className="rka-input-txt"
            disabled={false}
            id="angle"
            name="angle"
            onChange={[Function]}
            placeholder="angle (default: 90)"
          />
        </div>
        <div
          className="rka-form-group"
        >
          <label
            className="rka-label txt-cap"
            htmlFor="background_opacity"
          >
            background opacity
          </label>
          <input
            className="rka-input-txt"
            disabled={false}
            id="background_opacity"
            name="background_opacity"
            onChange={[Function]}
          />
        </div>
      </div>
    </div>
  </div>,
  <div
    className="pa-md mb-xs pos-r bg-gray-lightest is-active bor-light"
  >
    <div>
      <button
        className="rka-close-icon"
        onClick={[Function]}
      />
      <button
        className="rka-edit-icon"
        onClick={[Function]}
      />
      <span
        className="rka-move-icon"
      />
      <h3
        className="rka-h3 txt-cap"
      >
        crop
      </h3>
    </div>
    <div
      className=""
    >
      <div
        className="rka-operation-content"
      >
        <div
          className="rka-form-group"
        >
          <label
            className="rka-label txt-cap"
            htmlFor="width"
          >
            width
          </label>
          <input
            className="rka-input-txt"
            disabled={false}
            id="width"
            name="width"
            onChange={[Function]}
          />
        </div>
        <div
          className="rka-form-group"
        >
          <label
            className="rka-label txt-cap"
            htmlFor="height"
          >
            height
          </label>
          <input
            className="rka-input-txt"
            disabled={false}
            id="height"
            name="height"
            onChange={[Function]}
          />
        </div>
        <div
          className="rka-form-group"
        >
          <label
            className="rka-label txt-cap"
            htmlFor="anchor"
          >
            anchor
          </label>
          <input
            className="rka-input-txt"
            disabled={false}
            id="anchor"
            name="anchor"
            onChange={[Function]}
            placeholder="anchor (default: left)"
          />
        </div>
      </div>
    </div>
  </div>,
  <div
    className="pa-md bor-light mt-md"
  >
    <h3
      className="rka-h3 mb-md"
    >
      New operation
    </h3>
    <div
      className="rka-form-group"
    >
      <select
        className="rka-select"
        onChange={[Function]}
        value=""
      >
        <option
          value="crop"
        >
          crop
        </option>
        <option
          value="rotate"
        >
          rotate
        </option>
      </select>
    </div>
    <button
      className="rka-button rka-button-brand rka-button-sm"
      onClick={[Function]}
    >
      Add operation
    </button>
  </div>,
]
`;

exports[`OperationList does render with minimal props 1`] = `
Array [
  <h3
    className="rka-h2 mv-md"
  >
    Operations
  </h3>,
  <div
    className="pa-md bor-light mt-md"
  >
    <h3
      className="rka-h3 mb-md"
    >
      New operation
    </h3>
    <div
      className="rka-form-group"
    >
      <select
        className="rka-select"
        onChange={[Function]}
        value=""
      />
    </div>
    <button
      className="rka-button rka-button-brand rka-button-sm"
      onClick={[Function]}
    >
      Add operation
    </button>
  </div>,
]
`;
